nilearn/signal.py

```diff
@@ -452,8 +453,22 @@ def clean(signals, detrend=True, standardize=True, confounds=None,
 
         confounds = _ensure_float(confounds)
         confounds = _standardize(confounds, normalize=True, detrend=detrend)
-        Q = linalg.qr(confounds, mode='economic')[0]
-        signals -= np.dot(Q, np.dot(Q.T, signals))
+
+        if (LooseVersion(scipy.__version__) > LooseVersion('0.9.0')):
+            # Pivoting in qr decomposition was added in scipy 0.10
+            Q, R, _ = linalg.qr(confounds, mode='economic', pivoting=True)
+            Q = Q[:, np.abs(np.diag(R)) > np.finfo(np.float).eps * 100.]
+            signals -= Q.dot(Q.T).dot(signals)
+        else:
+            Q, R = linalg.qr(confounds, mode='economic')
+            non_null_diag = np.abs(np.diag(R)) > np.finfo(np.float).eps * 100.
+            if np.all(non_null_diag):
+                signals -= Q.dot(Q.T).dot(signals)
+            elif np.any(non_null_diag):
+                R = R[:, non_null_diag]
+                confounds = confounds[:, non_null_diag]
+                inv = scipy.linalg.inv(np.dot(R.T, R))
+                signals -= confounds.dot(inv).dot(confounds.T).dot(signals)
```

Comment:
that part is fine: What we want is the projector `confounds.dot(np.linalg.pinv(confounds)).dot(signals)`, which is `confounds.dot(pinv(confounds.T.dot(confounds))).dot(confounds.T).dot(signals)`, which amounts to what is written, because `confounds = Q.dot(R)` and `confounds.T.dot(confounds) = R.T.dot(Q.T).dot(Q).dot(R)` and `Q` disappears because it is orthogonal.

But as mentioned in an outside comment, this would have been handled by `confounds.dot(pinv(confounds).dot(signal))` also. The only question is where `pinv` cuts off singular values. AFAIK this is at `1e-15`.